lib/ssz_ex.ex

```diff
@@ -8,22 +8,222 @@ defmodule LambdaEthereumConsensus.SszEx do
   def encode(value, {:int, size}), do: encode_int(value, size)
   def encode(value, :bool), do: encode_bool(value)
 
+  def encode(list, {:list, basic_type, size}) do
+    if variable_size?(basic_type),
+      do: encode_variable_size_list(list, basic_type, size),
+      else: encode_fixed_size_list(list, basic_type, size)
+  end
+
   def decode(binary, :bool), do: decode_bool(binary)
   def decode(binary, {:int, size}), do: decode_uint(binary, size)
 
+  def decode(binary, {:list, basic_type, size}) do
+    if variable_size?(basic_type),
+      do: decode_variable_list(binary, basic_type, size),
+      else: decode_list(binary, basic_type, size)
+  end
+
   #################
   ### Private functions
   #################
-  defp encode_int(value, size) when is_integer(value), do: <<value::size(size)-little>>
+  @bytes_per_length_offset 4
+  @bits_per_byte 8
+
+  defp encode_int(value, size) when is_integer(value), do: {:ok, <<value::size(size)-little>>}
+  defp encode_bool(true), do: {:ok, "\x01"}
+  defp encode_bool(false), do: {:ok, "\x00"}
 
   defp decode_uint(binary, size) do
     <<element::integer-size(size)-little, _rest::bitstring>> = binary
-    element
+    {:ok, element}
+  end
+
+  defp decode_bool("\x01"), do: {:ok, true}
+  defp decode_bool("\x00"), do: {:ok, false}
+
+  defp encode_fixed_size_list(list, _basic_type, max_size) when length(list) > max_size,
+    do: {:error, "invalid max_size of list"}
+
+  defp encode_fixed_size_list(list, basic_type, _size) when is_list(list) do
+    list
+    |> Enum.map(&encode(&1, basic_type))
+    |> flatten_results_by(&Enum.join/1)
+  end
+
+  defp encode_variable_size_list(list, _basic_type, max_size) when length(list) > max_size,
+    do: {:error, "invalid max_size of list"}
+
+  defp encode_variable_size_list(list, basic_type, _size) when is_list(list) do
+    fixed_lengths = @bytes_per_length_offset * length(list)
+
+    with {:ok, {encoded_variable_parts, variable_offsets_list, total_byte_size}} <-
+           encode_variable_parts(list, basic_type) do
+      if fixed_lengths + total_byte_size <
+           2 ** (@bytes_per_length_offset * @bits_per_byte) do
+        {variable_offsets, _} =
+          Enum.reduce(variable_offsets_list, {[], 0}, fn element, {res, acc} ->
+            sum = fixed_lengths + acc
+            {[sum | res], element + acc}
+          end)
+
+        with {:ok, encoded_variable_offsets} <-
+               variable_offsets
+               |> Enum.reverse()
+               |> Enum.map(&encode(&1, {:int, 32}))
+               |> flatten_results() do
+          (encoded_variable_offsets ++ encoded_variable_parts)
+          |> :binary.list_to_bin()
+          |> then(&{:ok, &1})
+        end
+      else
+        {:error, "invalid lengths"}
+      end
```

Comment:
Nit. We can extract the `if` into a helper function:

```elixir
defp check_length(fixed_lengths, total_byte_size)
  when fixed_lengths + total_byte_size < 2 ** (@bytes_per_length_offset * @bits_per_byte),
  do: :ok
# handle the other case
```

Then you can merge both `with`s easily:

```elixir
with {:ok, ...} <- encode...,
  :ok <- check_length(...),
  {variable_offsets, _} = ...,
  {:ok, ...} <- variable_offsets... do
    ...
end
```